mailpoet/lib/Automation/Engine/Hooks.php

```diff
@@ -2,10 +2,33 @@
 
 namespace MailPoet\Automation\Engine;
 
+use MailPoet\Automation\Engine\Data\Step;
+use MailPoet\Automation\Engine\Data\Workflow;
+
 class Hooks {
+  /** @var WordPress */
+  private $wordPress;
+
+  public function __construct(
+    WordPress $wordPress
+  ) {
+    $this->wordPress = $wordPress;
+  }
+
   public const INITIALIZE = 'mailpoet/automation/initialize';
   public const API_INITIALIZE = 'mailpoet/automation/api/initialize';
   public const STEP_RUNNER_INITIALIZE = 'mailpoet/automation/step_runner/initialize';
   public const TRIGGER = 'mailpoet/automation/trigger';
   public const WORKFLOW_STEP = 'mailpoet/automation/workflow/step';
+
+  public const WORKFLOW_BEFORE_SAVE = 'mailpoet/automation/workflow/before_save';
+  public const WORKFLOW_STEP_BEFORE_SAVE = 'mailpoet/automation/workflow/step/before_save';
```

Comment:
Good point. There's one thing I'm not sure about — should we use filters or actions or both? Or should we have actions for now and later convert them to filters if we allow modifying some of the data?

Thinking about it, I agree with the `_ACTION` suffix — that way, we can keep both actions and filters in the same class and choose action vs. filter freely. I already had `Hook naming` on the checklist (to review it before MVP) so I now changed it to `Hook naming (review the naming and actions vs. filters)`. It will be probably better to review separately than to try to solve it within this PR.

Reply:
Hmm, I'm looking at this, and I'm not sure. WP doesn't reflect action vs. filter in the name (https://adambrown.info/p/wp_hooks/hook). I know we can have it only in the constants but wouldn't it prevent us from simply switching from action to a filter in the future? Or I guess such a switch is not without a BC break anyway because a filter must always return something? 🤔

Reply:
> should we use filters or actions or both? Or should we have actions for now and later convert them to filters if we allow modifying some of the data?
I think we'll ultimately just need to play this by ear. My gut reaction is that we should use actions where possible. We can even allow modification of data via actions since objects will be passed by reference, and we could expose public methods on objects that could be used to modify data if we wanted to, though that would arguably be moving a little too far away from WP standards. Filters seem to be more prone to problems, since it's possible for a callback to return a completely different type, or return nothing at all.

As for including `action` or `filter` in the names, I do think that's probably overkill. It could be a tiny nice-to-have to use a suffix for the constants, but in any case I don't think we need to make any changes right away.

Reply:
> My gut reaction is that we should use actions where possible.
> Filters seem to be more prone to problems, since it's possible for a callback to return a completely different type, or return nothing at all.

Good point, I haven't thought about that. WP filters indeed allow absolutely anything, which is problematic. I initially almost started doing a custom event publisher/listener service but then I remembered it makes no sense — when we're in WP we probably should go with actions and filters.

I'm just thinking now to what extent we can wrap WP hooks and "type" them. That, of course, won't prevent anyone from using them directly, but we could have a higher-level safer way of using them. For now, I added a few helper methods (as you saw in this PR) but maybe that's all subject to change and improvement. Definitely leaving this open to further exploration.